// node-red-contrib-home-assistant-websocket/dist/const.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SERVER_ADD = exports.ZONE_LEAVE = exports.ZONE_ENTER_OR_LEAVE = exports.ZONE_ENTER = exports.TYPEDINPUT_STR = exports.TYPEDINPUT_NUM = exports.TYPEDINPUT_NONE = exports.TYPEDINPUT_JSONATA = exports.TYPEDINPUT_JSON = exports.TYPEDINPUT_ENV = exports.TYPEDINPUT_BOOL = exports.TYPEDINPUT_DATE = exports.TYPEDINPUT_GLOBAL = exports.TYPEDINPUT_FLOW = exports.TYPEDINPUT_MSG = exports.STATE_RUNNING = exports.STATE_ERROR = exports.STATE_DISCONNECTED = exports.STATE_CONNECTED = exports.STATE_CONNECTING = exports.INTEGRATION_UNLOADED = exports.INTEGRATION_NOT_LOADED = exports.INTEGRATION_LOADED = exports.INTEGRATION_EVENT = exports.HA_EVENTS = exports.HA_EVENT_TAG_SCANNED = exports.HA_EVENT_STATE_CHANGED = exports.HA_EVENT_INTEGRATION = exports.HA_EVENT_SERVICES_UPDATED = exports.HA_EVENT_REGISTRY_UPDATED = exports.HA_EVENT_DEVICE_REGISTRY_UPDATED = exports.HA_EVENT_AREA_REGISTRY_UPDATED = exports.HA_CLIENT_READY = exports.ENTITY_DEVICE_TRIGGER = exports.ENTITY_SWITCH = void 0;
exports.ENTITY_SWITCH = 'switch';
exports.ENTITY_DEVICE_TRIGGER = 'device_trigger';
exports.HA_CLIENT_READY = 'ha_client:ready';
exports.HA_EVENT_AREA_REGISTRY_UPDATED = 'areas_updated';
exports.HA_EVENT_DEVICE_REGISTRY_UPDATED = 'devices_updated';
exports.HA_EVENT_REGISTRY_UPDATED = 'registry_updated';
exports.HA_EVENT_SERVICES_UPDATED = 'services_updated';
exports.HA_EVENT_INTEGRATION = 'nodered';
exports.HA_EVENT_STATE_CHANGED = 'state_changed';
exports.HA_EVENT_TAG_SCANNED = 'tag_scanned';
exports.HA_EVENTS = 'ha_events';
exports.INTEGRATION_EVENT = 'integration';
exports.INTEGRATION_LOADED = 'loaded';
exports.INTEGRATION_NOT_LOADED = 'notloaded';
exports.INTEGRATION_UNLOADED = 'unloaded';
exports.STATE_CONNECTING = 0;
exports.STATE_CONNECTED = 1;
exports.STATE_DISCONNECTED = 2;
exports.STATE_ERROR = 3;
exports.STATE_RUNNING = 4;
exports.TYPEDINPUT_MSG = 'msg';
exports.TYPEDINPUT_FLOW = 'flow';
exports.TYPEDINPUT_GLOBAL = 'global';
exports.TYPEDINPUT_DATE = 'date';
exports.TYPEDINPUT_BOOL = 'bool';
exports.TYPEDINPUT_ENV = 'env';
exports.TYPEDINPUT_JSON = 'json';
exports.TYPEDINPUT_JSONATA = 'jsonata';
exports.TYPEDINPUT_NONE = 'none';
exports.TYPEDINPUT_NUM = 'num';
exports.TYPEDINPUT_STR = 'str';
exports.ZONE_ENTER = 'enter';
exports.ZONE_ENTER_OR_LEAVE = 'enter_leave';
exports.ZONE_LEAVE = 'leave';
exports.SERVER_ADD = '_ADD_';
